pkg/core/klinedriver/tick_klinedriver.go

```diff
@@ -0,0 +1,180 @@
+	return &TickKLineDriver{
+		symbol:        symbol,
+		tickInterval:  tickInterval,
+		intervals:     make(map[types.Interval]struct{}),
+		intervalSlice: []types.Interval{},
+		builder:       NewKLineBuilder(symbol),
+		running:       false,
+		tradesBuffer:  []*types.Trade{},
+	}
+}
+
+func (t *TickKLineDriver) SetKLineEmitter(emitter KLineEmitter) {
+	t.kLineEmitter = emitter
+}
+
+// SetRunning sets the running state of the driver.
+// As for backtesting, use this method to set the start time of the driver.
+func (t *TickKLineDriver) SetRunning(startTime time.Time) {
+	// add intervals
+	for interval := range t.intervals {
+		// initialize the kline with the truncated start time
+		startTimeTruncated := startTime.Truncate(interval.Duration())
+		t.builder.AddInterval(interval, types.Time(startTimeTruncated))
+	}
+	t.running = true
+}
+
+func (t *TickKLineDriver) Subscribe(interval types.Interval) error {
+	if interval.Duration() < t.tickInterval.Duration() {
+		return fmt.Errorf("interval %s can't be smaller than tick interval %s", interval, t.tickInterval)
+	}
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	if _, ok := t.intervals[interval]; ok {
+		return nil
+	}
+	t.intervals[interval] = struct{}{}
+	t.intervalSlice = append(t.intervalSlice, interval)
+	return nil
+}
+
+// AddTrade adds a trade to the kline driver.
+func (t *TickKLineDriver) AddTrade(trade types.Trade) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	// set trade time location
+	if t.tradeTimeLocation == nil {
+		t.tradeTimeLocation = trade.Time.Time().Location()
+	}
+
+	// it's not running yet, buffer the trade
+	if !t.running {
+		t.tradesBuffer = append(t.tradesBuffer, &trade)
+		return
+	}
+	// it's running, add buffered trades first
+	if len(t.tradesBuffer) > 0 {
+		for _, bufferTrade := range t.tradesBuffer {
+			t.builder.AddTrade(bufferTrade)
+		}
+		// empty the buffer
+		t.tradesBuffer = []*types.Trade{}
+	}
+	t.builder.AddTrade(&trade)
+}
+
+// Run starts the kline driver.
+// The driver can be only run once. Panic otherwise.
+func (t *TickKLineDriver) Run(ctx context.Context) {
+	go t.run(ctx)
+}
+
+func (t *TickKLineDriver) run(ctx context.Context) {
+	if t.running {
+		panic("can not run kline driver twice")
+	}
+
+	t.mu.Lock()
+	// start ticker
+	log.Debugf("start kline driver: %s(%s) %v", t.symbol, t.tickInterval, t.intervalSlice)
+	ticker := time.NewTicker(t.tickInterval.Duration())
+	defer ticker.Stop()
+
+	// set running state
+	t.SetRunning(time.Now())
+
+	// add buffered trades if any
+	for _, trade := range t.tradesBuffer {
+		t.builder.AddTrade(trade)
+	}
+	t.mu.Unlock()
+
+	// update loop
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case tickTime := <-ticker.C:
+			t.ProcessTick(tickTime)
+		}
+	}
+}
+
+// ProcessTick processes the tick time and emits kline events.
+func (t *TickKLineDriver) ProcessTick(tickTime time.Time) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	if t.tradeTimeLocation == nil {
```

Comment:
tradeTimeLocation should not be necessary? since we don't set timezone for the trades

Reply:
Since drawing the K-line varies depending on the time zone, shouldn’t time.Location theoretically be a parameter here, instead of being passed in from the trade?

Reply:
I see.

Let's fix the time zone as `UTC` for all klines for now.
If we need to support different time zone in the future, it's not hard to add the time zone as parameter afterward.

Reply:
okay